src/filters/list.js

```diff
@@ -0,0 +1,207 @@
+/**
+ * @license Copyright (c) 2003-2018, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+/**
+ * @module paste-from-office/filters/list
+ */
+
+import Element from '@ckeditor/ckeditor5-engine/src/view/element';
+import Matcher from '@ckeditor/ckeditor5-engine/src/view/matcher';
+import Range from '@ckeditor/ckeditor5-engine/src/view/range';
+import UpcastWriter from '@ckeditor/ckeditor5-engine/src/view/upcastwriter';
+
+/**
+ * Transforms Word specific list-like elements to the semantic HTML lists.
+ *
+ * Lists in Word are represented by block elements with special attributes like:
+ *
+ *		<p class=MsoListParagraphCxSpFirst style='mso-list:l1 level1 lfo1'>...</p> // Paragraph based list.
+ *		<h1 style='mso-list:l0 level1 lfo1'>...</h1> // Heading 1 based list.
+ *
+ * @param {module:engine/view/documentfragment~DocumentFragment} documentFragment The view structure which to transform.
+ * @param {String} stylesString Styles from which list-like elements styling will be extracted.
+ */
+export function transformListItemLikeElementsIntoLists( documentFragment, stylesString ) {
+	if ( !documentFragment.childCount ) {
+		return;
+	}
+
+	const listLikeItems = findAllListItemLikeElements( documentFragment );
+
+	if ( listLikeItems.length ) {
+		const writer = new UpcastWriter();
+
+		let currentList = null;
+
+		for ( let i = 0; i < listLikeItems.length; i++ ) {
+			if ( !currentList || listLikeItems[ i - 1 ].id !== listLikeItems[ i ].id ) {
+				const listStyle = findListStyle( listLikeItems[ i ], stylesString );
+				currentList = insertEmptyList( listStyle, listLikeItems[ i ].element, writer );
+			}
+
+			const listItem = transformElementIntoListItem( listLikeItems[ i ].element, writer );
```

Comment:
I had to read deep into the code to understand why we don't remove the old element if this function returns a new element. You need to know about the internals of this function to figure this out.

Reply:
In short – I wouldn't use `rename()` in such a context for its complicated nature. I'd rather explicitly create a new element and remove the old one.

Reply:
But this is a bit too much work now, so I'm skipping it.